lib/ex_unit/lib/ex_unit/capture_server.ex

```diff
@@ -30,28 +34,25 @@ defmodule ExUnit.CaptureServer do
 
   def init(:ok) do
     state = %{
-      devices: {%{}, %{}},
+      devices: %{},
       log_captures: %{},
       log_status: nil
     }
 
     {:ok, state}
   end
 
-  def handle_call({:device_capture_on, name, pid}, _from, config) do
-    {names, refs} = config.devices
+  def handle_call({:device_capture_on, name, encoding, input}, {caller, _}, config) do
+    capture_device(name, encoding, input, config, caller)
+  end
 
-    if Map.has_key?(names, name) do
-      {:reply, {:error, :already_captured}, config}
-    else
-      orig_pid = Process.whereis(name)
-      Process.unregister(name)
-      Process.register(pid, name)
-      ref = Process.monitor(pid)
-      refs = Map.put(refs, ref, {name, orig_pid})
-      names = Map.put(names, name, true)
-      {:reply, {:ok, ref}, %{config | devices: {names, refs}}}
-    end
+  def handle_call({:device_output, ref}, _from, %{devices: devices} = config) do
+    {_, device} = device_by_ref(devices, ref)
```

Comment:
Please pass the name as argument and if you convert `refs` to a map, than this can be:

    byte_offset = config.devices[name].refs[ref]

Which should be much simpler than traversing looking for refs.

Reply:
Alternatively, you can keep another map, of refs to device name. This way you can use the reverse map lookup here, and also on `DOWN` and `device_capture_off`.